internal/pkg/addon/package.go

```diff
@@ -0,0 +1,243 @@
+	// after uploading an asset to S3. If this and BucketNameProperty are empty,
+	// a submap will not be created and an S3 location URI will replace value of Property.
+	ObjectKeyProperty string
+
+	// ForceZip will force a zip file to be created even if the given file URI
+	// points to a file. Directories are always zipped.
+	ForceZip bool
+}
+
+// resourcePackageConfig maps a CloudFormation resource type to configuration
+// for how to transform it's properties.
+//
+// TODO(dnrnd) AWS::Include.Location
+var resourcePackageConfig = map[string][]packagePropertyConfig{
+	"AWS::ApiGateway::RestApi": {
+		{
+			Property:           []string{"BodyS3Location"},
+			BucketNameProperty: "Bucket",
+			ObjectKeyProperty:  "Key",
+		},
+	},
+	"AWS::Lambda::Function": {
+		{
+			Property:           []string{"Code"},
+			BucketNameProperty: "S3Bucket",
+			ObjectKeyProperty:  "S3Key",
+			ForceZip:           true,
+		},
+	},
+	"AWS::Lambda::LayerVersion": {
+		{
+			Property:           []string{"Content"},
+			BucketNameProperty: "S3Bucket",
+			ObjectKeyProperty:  "S3Key",
+			ForceZip:           true,
+		},
+	},
+	"AWS::CloudFormation::Stack": {
+		{
+			// This implementation does not recursively package
+			// the local template pointed to by TemplateURL.
+			Property: []string{"TemplateURL"},
+		},
+	},
+	"AWS::AppSync::GraphQLSchema": {
+		{
+			Property: []string{"DefinitionS3Location"},
+		},
+	},
+	"AWS::AppSync::Resolver": {
+		{
+			Property: []string{"RequestMappingTemplateS3Location"},
+		},
+		{
+			Property: []string{"ResponseMappingTemplateS3Location"},
+		},
+	},
+	"AWS::AppSync::FunctionConfiguration": {
+		{
+			Property: []string{"RequestMappingTemplateS3Location"},
+		},
+		{
+			Property: []string{"ResponseMappingTemplateS3Location"},
+		},
+	},
+	"AWS::ElasticBeanstalk::ApplicationVersion": {
+		{
+			Property:           []string{"SourceBundle"},
+			BucketNameProperty: "S3Bucket",
+			ObjectKeyProperty:  "S3Key",
+		},
+	},
+	"AWS::Glue::Job": {
+		{
+			Property: []string{"Command", "ScriptLocation"},
+		},
+	},
+	"AWS::StepFunctions::StateMachine": {
+		{
+			Property:           []string{"DefinitionS3Location"},
+			BucketNameProperty: "Bucket",
+			ObjectKeyProperty:  "Key",
+		},
+	},
+	"AWS::CodeCommit::Repository": {
+		{
+			Property:           []string{"Code", "S3"},
+			BucketNameProperty: "Bucket",
+			ObjectKeyProperty:  "Key",
+			ForceZip:           true,
+		},
+	},
+}
+
+func (a *Addons) packageTemplate(tmpl *cfnTemplate) (*cfnTemplate, error) {
+	resources := mappingNode(&tmpl.Resources)
+
+	for name, node := range resources {
+		resType := yamlMapGet(node, "Type").Value
+		confs, ok := resourcePackageConfig[resType]
+		if !ok {
+			continue
+		}
+
+		props := yamlMapGet(node, "Properties")
+		for _, conf := range confs {
+			if err := a.packageProperty(props, conf); err != nil {
+				return nil, fmt.Errorf("transform property %s property for %s: %w", name, conf.Property, err)
```

Comment:
What do you think of this rewording?
```suggestion
				return nil, fmt.Errorf("package property %q of %q: %w",  strings.Join(conf.Property, "."), name, err)
```